exporters/pelorus/utils/nested.py

```diff
@@ -0,0 +1,191 @@
+class _NoDefault(enum.Enum):
+    NO_DEFAULT = enum.auto()
+
+
+@overload
+def get_nested(
+    root: Mapping[str, T],
+    path: Union[Sequence[str], str],
+    *,
+    name: Optional[str] = None,
+) -> T:
+    ...
+
+
+@overload
+def get_nested(
+    root: Mapping[str, T],
+    path: Union[Sequence[str], str],
+    *,
+    default: U,
+    name: Optional[str] = None,
+) -> Union[T, U]:
+    ...
+
+
+def get_nested(
+    root: Mapping[str, T],
+    path: Union[Sequence[str], str],
+    *,
+    default: Union[U, Literal[_NoDefault.NO_DEFAULT]] = _NoDefault.NO_DEFAULT,
+    name: Optional[str] = None,
+) -> Union[T, U]:
+    """
+    `get_nested` helps you safely traverse a deeply nested object that is indexable.
+    If `TypeError`, `KeyError`, or `IndexError` are thrown, then `default` will be returned.
+    If `default` is not given, a `MissingAttributeError` will be thrown,
+    which includes information about where in the path things went wrong, and a human-readable name (if included).
+
+    You may specify the path as either a list of keys, or a single string.
+    The string will be split on '.' so you can emulate the nested attribute lookup `ResourceField`
+    would offer.
+
+    A `name` for the item, if specified, makes the error message in the exception more useful.
+
+    Kubernetes API items often are deeply nested, with any number of fields that could be absent.
+    When using an `openshift.dynamic.ResourceField`, it will turn attribute accesses into
+    dictionary accesses. Normally, a deeply nested access like item.status.ref.foo.bar has four different spots
+    you could get an `AttributeError`. With a `ResourceField`, there are actually only three, since `item.status`
+    will return `None` if `status` is absent, but `None` will not have a `ref` field, leading to an
+    AttributeError.
+    """
+    item = root
+    path = split_path(path)
+
+    for i, key in enumerate(path):
+        try:
+            item = item[key]  # type: ignore
+        except (TypeError, IndexError, KeyError) as e:
+            if default is not _NoDefault.NO_DEFAULT:
+                return default
+
+            raise BadAttributePathError(
+                root=root,
+                path=path,
+                path_slice=slice(i),
+                value=item,
+                root_name=name,
+            ) from e
+
+    return item
+
+
+def split_path(path: Union[str, Sequence[str]]) -> Sequence[str]:
+    """
+    Idempotently split a path for use in nested access.
+    """
+    if isinstance(path, str):
+        # `if part` filters out leading dot (or accidental double dots, technically)
+        return tuple(part for part in path.split(".") if part)
+    else:
+        return path
+
+
+def format_path(path: Sequence[str]) -> str:
+    """
+    Format a path for readability.
+    >>> assert format_path("foo.bar".split(".")) == "foo.bar"
+    >>> assert format_path(["foo", "has.dots", "bar"]) == "foo[has.dots].bar"
```

Comment:
this doctests are breaking. Do we run them in the CI?

Reply:
since there are some, I would remove them from the doctrings and add to the tests

Reply:
Fixed in the architecture docs PR